src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1122,7 +1122,7 @@ trait Namers extends MethodSynthesis {
           val leg = legacy.toString
           val help = if (pts != leg) s" instead of $leg" else ""
           val msg = s"under -Xsource:3-cross, the inferred type changes to $pts$help"
-          val src = currentUnit.source
+          val src = tree.pos.source
```

Comment:
Ah I didn't realize this PR also fixes https://github.com/scala/bug/issues/12941, I guess that fix is here? I'll add a test.